src/rdkafka_cgrp.c

```diff
@@ -2434,6 +2435,280 @@ static void rd_kafka_cgrp_offset_commit_tmr_cb (rd_kafka_timers_t *rkts,
+
+        /* Verify partitions do not exist in the current assignment before
+         * making any changes to it to so the operation doesn't partially
+         * succeed. */
+        for (i = 0; rkcg->rkcg_assignment && i < partitions->cnt; i++) {
+                const rd_kafka_topic_partition_t *rktpar = &partitions->elems[i];
+                if (rd_kafka_topic_partition_list_find(rkcg->rkcg_assignment,
+                                                       rktpar->topic,
+                                                       rktpar->partition)) {
+                        return rd_kafka_error_new(
+                                RD_KAFKA_RESP_ERR__INVALID_ARG,
+                                "%s [%"PRId32"] is already part of the "
+                                "current assignment",
+                                rktpar->topic, rktpar->partition);
+                }
+        }
+
+        if (partitions->cnt == 0) {
+                rd_kafka_dbg(rkcg->rkcg_rk, CGRP|RD_KAFKA_DBG_CONSUMER,
+                        "ASSIGN", "Group \"%s\": adding 0 partitions to "
+                        "existing assignment of %d partitions in join state "
+                        "%s (nothing to do)", rkcg->rkcg_group_id->str,
+                        !rkcg->rkcg_assignment ? 0
+                        : rkcg->rkcg_assignment->cnt,
+                        rd_kafka_cgrp_join_state_names[rkcg->rkcg_join_state]);
+                return NULL;
+        }
+
+        rd_kafka_cgrp_version_new_barrier(rkcg);
+
+        rd_kafka_dbg(rkcg->rkcg_rk, CGRP|RD_KAFKA_DBG_CONSUMER, "ASSIGN",
+                "Group \"%s\": adding %d partitions to existing assignment "
+                "of %d partitions in join state %s (v%"PRId32")",
+                rkcg->rkcg_group_id->str, partitions->cnt,
+                !rkcg->rkcg_assignment ? 0 : rkcg->rkcg_assignment->cnt,
+                rd_kafka_cgrp_join_state_names[rkcg->rkcg_join_state],
+                rkcg->rkcg_version);
+
+        if (!rkcg->rkcg_assignment)
+                rkcg->rkcg_assignment = rd_kafka_topic_partition_list_new(
+                                                partitions->cnt);
+
+        /* Add and mark partition(s) as desired */
+        for (i = 0; i < partitions->cnt; i++) {
+                rd_kafka_toppar_t *rktp;
+                const rd_kafka_topic_partition_t *rktpar =
+                        &partitions->elems[i];
+
+                /* This marks the partition as desired if it's not already. */
+                rktp = rd_kafka_toppar_get2(rkcg->rkcg_rk,
+                                            rktpar->topic,
+                                            rktpar->partition,
+                                            0/*no-ua*/, 1/*create-on-miss*/);
+
+                if (!rktp)
+                        /* LOG_ERR already emitted by rd_kafka_toppar_get2 */
+                        continue;
+
+                rd_kafka_toppar_keep(rktp);
```

Comment:
Don't think this is needed since get2 increases the refcount for the returned rktp.

Reply:
i can assure you it is necessary based on refcount problems that occur if it's not here... i arrived  at this basically by starting with cgrp_assign/unassign methods and changing them for incremental updates. I will try to understand clearly why this is necessary and write a further note.

Reply:
Ah, right, one refcount (from get2()) goes on partitions, and the other (from keep()) goes on rkcg_assignment.
I think it is better to change add0() to keep() the rktp if it is specified and remove the keep from here.
Also update the one place where add0 is called with a keep().

Reply:
sounds good. but two places? `rd_kafka_toppar_offset_fetch` and `rd_kafka_topic_partition_add_copy`